docs: update releases.md per SRU review feedback on branch structuring (#1645)
canonical/ubuntu-advantage-client: RELEASES.md

| Build target | Version Format |
| -------- | -------- |
| Devel series upstream release | `XX.YY` |
| Devel series bugfix release | `XX.YY.Z~ubuntu1`|
| Stable series release | `XX.YY~ubuntu1~18.04.1`|
| [Daily Build Recipe](https://code.launchpad.net/~canonical-server/+recipe/ua-client-daily) | `XX.YY+<revtime>-g<commitish>~ubuntu1~18.04.1` |
| Ubuntu PRO series release | binary copies of Daily PPA |

## Supported upgrade use-cases based on version formats

> | Upgrade path | Version diffs |

Reviewer: This table isn't rendering in github, I think because its missing this line
```suggestion
| Upgrade path | Version diffs |
| -------- | -------- |
```

| LTS to LTS | `20.3~ubuntu1~14.04.1 -> 20.3~ubuntu1~16.04.1` |
| LTS to Daily PPA | `20.3~ubuntu1~14.04.1 -> 20.3+202004011202~ubuntu1~14.04.1` |
| Ubuntu PRO to latest <series>-updates | `20.3+202004011202~ubuntu1~14.04.1` -> `20.4~ubuntu1~14.04.1` |
| Ubuntu PRO to Daily PPA | `20.3+202004011202~ubuntu1~14.04.1` -> `20.4+202004021202~ubuntu1~14.04.1` |

## Devel Release Process

> Below is the procedure used to release ubuntu-advantage-client to an Ubuntu series, it assumes we already have an annotated tag via `git tag -a 27.1` on the `release-27` branch which has not yet been uploaded to `pkg/ubuntu/devel`:

Reviewer: There's another related assumption in an comment in the middle of the list. It would help me to see them together up front in the rendered version. Maybe something like this:
```suggestion
Below is the procedure used to release ubuntu-advantage-client to an Ubuntu series. It makes the following assumptions:
* The current ubuntu devel release is impish
* We are releasing version 27.1
* The previous release was 27.0.2
* The commits for 27.0.2 and 27.1 on https://github.com/canonical/ubuntu-advantage-client are tagged with `27.0.2` and `27.1` respectively.
  * These tags should be on the correct commits on the `release-27` branch
  * `27.1` which is the tag of the release we are making, has not yet been uploaded to `pkg/ubuntu/devel`
```

 1. git-ubuntu clone ubuntu-advantage-tools; cd ubuntu-advantage-tools
 2. git remote add upstream git@github.com:canonical/ubuntu-advantage-client.git
 3. git fetch upstream

>  \# Assume last previous released tag on `release-27` branch was `27.0.2`

Reviewer: If we accept my suggestion above, we can remove this line
```suggestion
```

 4. git rebase --onto pkg/ubuntu/devel 27.0.2 27.1
 5. git checkout -B upload-27.1-impish # to create a new local branch name based on your detached branch in step 4
 6. git push `<your_launchpad_user>` upload-27.1-impish
 7. Create a PPA for upload reviewers at `https://launchpad.net/~<YOUR_LP_USER>/+activate-ppa`
 8. Push to your test PPA so upload reviewers can install easily test packages easily during review
 ```bash
    debuild -S
    dput  ppa:<YOUR_LAUNCHPAD_USER>/ua-client-uploads ./ubuntu-advantage-tools_*_source.changes
```

>  9. Create a merge proposal such as [27.0.2 PR](https://code.launchpad.net/~chad.smith/ubuntu/+source/ubuntu-advantage-tools/+git/ubuntu-advantage-tools/+merge/402459) which targets `ubuntu/devel`

Reviewer: My attempt at rewording this one
```suggestion
 9. Create a merge proposal for 27.1 which targets `ubuntu/devel`
     * For an example, see the [27.0.2 merge proposal](https://code.launchpad.net/~chad.smith/ubuntu/+source/ubuntu-advantage-tools/+git/ubuntu-advantage-tools/+merge/402459)
```

 10. Add 2 review slots for `canonical-server` and `canonical-server-core-reviewers`

>  11. **For SRUs:** Create an "upload a new version bug" in https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bugs like [27.0](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1926361) or [20.3](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1869980)

Author: Done separated and added more info for SRUs

>  11. **For SRUs:** Create an "upload a new version bug" in https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bugs like [27.0](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1926361) or [20.3](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1869980)
>  12. Describe the need, provide testing PPA and describe test instructions

Reviewer: Suggest breaking the instructions and examples into sub-bullets (not actually sure if sub-bullets work though in gihub markdown)
```suggestion
 11. **For SRUs:** Create an "upload a new version bug" in https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bugs
     * Describe the need, provide testing PPA and describe test instructions
     * For examples, see [27.0](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1926361) or [20.3](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1869980)
```

 13. Create a PR for each target series based off the latest `ubuntu/devel` accepted upload
 ```bash
 UA_VERSION=27.1
 SRU_BUG=TODO_FROM_STEP_11
 rm -rf ../out
 for release in xenial bionic focal groovy hirsute do
  git checkout release-${UA_VERSION}-impish -B upload-${UA_VERSION}-$release
  case "${release}" in
      xenial) version=${UA_VERSION}~16.04.1;;
      bionic) version=${UA_VERSION}~18.04.1;;
      focal) version=${UA_VERSION}~20.04.1;;
      groovy) version=${UA_VERSION}~20.10.1;;
      hirsute) version=${UA_VERSION}~21.04.1;;
  esac
  dch -v ${version} -D ${release} Backport new upstream release: (LP: #${SRU_BUG}) to $release
  git commit -m 'changelog backport to ${release}' debian/changelog
  git push <your_launchpad_username> upload-${UA_VERSION}-$release

>   build-package # to create release-specific dsc files used to upload in step 16

Author: I'm hoping if this is agreeable, is useful and works for the next SRU uploader and review cycle, then we can automate it as a script and reduce the docs.

 ```
 14. Create PRs for each SRU target release @ `https://code.launchpad.net/~<YOUR_LP_USER></YOUR_LAUNCHPAD_USER>/ubuntu/+source/ubuntu-advantage-tools/+git/ubuntu-advantage-tools/`. Make sure each PR targets the proper `ubuntu/$release-proposed` branch for each release.

>  15. Add both `canonical-server` and `canonical-server-core-reviewers` as review slots on the PR.

Author: > Maybe I'm remembering incorrectly, but shouldn't this be saying that each MP should target the impish branch?

You are correct here. this direction is concorrect, all branches should target release-27-impish. will fix.

> 
> Also, as annoying as it is that every platform uses different terms for PRs, we should try to use the appropriate term per platform, which means "merge proposal'/"MP" here I believe. 
Yeah, and even more broken and incorrect was my own mapping of proposal to the underlying SCM used. I was using PR for "git" based projects and "MP" for bazaar (because launchpad/bazaar were tightly coupled historically in my head).

So MP it is.

   * Describe the need, provide testing PPA and describe test instructions
   * For examples, see [27.0](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1926361) or [20.3](https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1869980)
 2. Create a PR for each target series based off your local `release-${UA_VERSION}-impish` branch:
 ```bash
 UA_VERSION=27.1
 SRU_BUG=TODO_FROM_STEP_11
 rm -rf ../out
 for release in xenial bionic focal groovy hirsute do
  git checkout release-${UA_VERSION}-impish -B upload-${UA_VERSION}-$release
  case "${release}" in
      xenial) version=${UA_VERSION}~16.04.1;;
      bionic) version=${UA_VERSION}~18.04.1;;
      focal) version=${UA_VERSION}~20.04.1;;
      groovy) version=${UA_VERSION}~20.10.1;;
      hirsute) version=${UA_VERSION}~21.04.1;;
  esac
  dch -v ${version} -D ${release} Backport new upstream release: (LP: #${SRU_BUG}) to $release
  git commit -m 'changelog backport to ${release}' debian/changelog
  git push <your_launchpad_username> upload-${UA_VERSION}-$release
  build-package # to create release-specific dsc files used to upload in step 16
 ```

>  3. Create merge proposals for each SRU target release @ `https://code.launchpad.net/~<YOUR_LP_USER></YOUR_LAUNCHPAD_USER>/ubuntu/+source/ubuntu-advantage-tools/+git/ubuntu-advantage-tools/`. Make sure each PR targets your `release-${UA_VERSION}-impish` branch.

Reviewer 2: The PR should target `impish` branch ? Shouldn't we target the devel branches for each release ?

Author: @lucasmoura this was decided as part of both the upload  and the SRU review is that it makes it easier to review given that we are syncing the exact same content (minus a changelog diff) into all supported Ubuntu releases. The upload reviewer will be able to handle the upload separately and not actually merge the releases into the impish branch as part of the review.

Reviewer 2: Oh okay, sorry, I forgot about that part. My mistake
